libftl/ingest.c

```diff
@@ -92,6 +92,111 @@ OS_THREAD_ROUTINE _ingest_get_rtt(void *data) {
     return 0;
 }
 
+ftl_status_t ftl_find_closest_available_ingest(const char* ingestHosts[], int ingestsCount, char* bestIngestHostComputed)
+{
+    ftl_ingest_t* ingestElements;
+    OS_THREAD_HANDLE *handles;
+    _tmp_ingest_thread_data_t *data;
+    
+    int i;
+
+    ftl_status_t ret_status = FTL_SUCCESS;
+    do {
+      if ((ingestElements = malloc(sizeof(ftl_ingest_t) * ingestsCount)) == NULL) {
+        ret_status = FTL_MALLOC_FAILURE;
+        break;
+      }
+
+      for (i = 0; i < ingestsCount; i++) {
+        size_t host_len = strlen(ingestHosts[i]) + 1;
+        if ((ingestElements[i].hostname = malloc(host_len)) == NULL) {
+          ret_status = FTL_MALLOC_FAILURE;
+          break;
+        }
+        strcpy_s(ingestElements[i].hostname, host_len, ingestHosts[i]);
+        ingestElements[i].rtt = 1000;
+        ingestElements[i].next = NULL;
+      }
+      if (ret_status != FTL_SUCCESS) {
+        break;
+      }
+
+      if ((handles = (OS_THREAD_HANDLE *)malloc(sizeof(OS_THREAD_HANDLE) * ingestsCount)) == NULL) {
+        ret_status = FTL_MALLOC_FAILURE;
+        break;
+      }
+
+      if ((data = (_tmp_ingest_thread_data_t *)malloc(sizeof(_tmp_ingest_thread_data_t) * ingestsCount)) == NULL) {
+        ret_status = FTL_MALLOC_FAILURE;
+        break;
+      }
+    } while (0);
+
+    // malloc failed, cleanup
+    if (ret_status != FTL_SUCCESS) {
+      if (ingestElements != NULL) {
+        for (i = 0; i < ingestsCount; i++) {
+          free(ingestElements[i].hostname);

```

Comment:
This also isn't safe, if you stopped midway through the loops some will be malloced and some won't be.

Reply:
are you mentioning that free will failed somehow because the hostname pointer is not initialized to NULL?

Reply:
Yeah, it's safe to call free with either NULL or an address that you malloced(). But it's not safe to call with anything else. So if you don't initialize all of the `ingestElements[i].hostname` to null, they will point at some random value, and thus calling free on them will be bad.

Reply:
Anyways, I think if you cmalloc the entire array memory when you create it you should be good. Then if you exit early whatever elements are malloced will be freeded, and the rest will be ignored.